src/object.rs

```diff
@@ -254,11 +256,17 @@ impl Object {
     /// # Ok(())
     /// # }
     /// ```
-    pub async fn range_read(&self, range: impl RangeBounds<u64>) -> Result<Vec<u8>> {
-        let op = OpRead::new(self.path(), range)?;
+    pub async fn range_read(&self, range: impl RangeBounds<u64> + Clone) -> Result<Vec<u8>> {
+        let op = OpRead::new(self.path(), range.clone())?;
         let s = self.acc.read(&op).await?;
 
-        let mut bs = Cursor::new(Vec::new());
+        let br = BytesRange::from(range);
```

Comment:
Maybe we can calculate the size before `OpRead::new` so that we don't need to clone it.

Reply:
We can implement

```rust
impl<T> From<&T> for BytesRange
where
    T: RangeBounds<u64>,
```

so we don't need to consume the `RangeBounds<u64>`.

Reply:
> We can implement

I tried this a lit, doesn't work.

How about adding a new function for `BytesRange`:

```rust
impl From<(Bound<u64>, Bound<u64>)> for BytesRange {
    fn from((start, end): (Bound<u64>, Bound<u64>)) -> Self {
        let offset = match start {
            Bound::Included(n) => Some(n),
            Bound::Excluded(n) => Some(n + 1),
            Bound::Unbounded => None,
        };
        let size = match end {
            Bound::Included(n) => Some(n + 1 - offset.unwrap_or_default()),
            Bound::Excluded(n) => Some(n - offset.unwrap_or_default()),
            Bound::Unbounded => None,
        };

        BytesRange(offset, size)
    }
}
```

`impl<T> From<T> for BytesRange` cloud also be updated via using bound instead.

Reply:
pretty solution!